src_c/scale2x.c

```diff
@@ -61,139 +59,36 @@ scale2x(SDL_Surface *src, SDL_Surface *dst)
     const int width = src->w;
     const int height = src->h;
 
-    switch (src->format->BytesPerPixel) {
-        case 1: {
-            Uint8 E0, E1, E2, E3, B, D, E, F, H;
-            for (looph = 0; looph < height; ++looph) {
-                for (loopw = 0; loopw < width; ++loopw) {
-                    B = *(Uint8 *)(srcpix + (MAX(0, looph - 1) * srcpitch) +
-                                   (1 * loopw));
-                    D = *(Uint8 *)(srcpix + (looph * srcpitch) +
-                                   (1 * MAX(0, loopw - 1)));
-                    E = *(Uint8 *)(srcpix + (looph * srcpitch) + (1 * loopw));
-                    F = *(Uint8 *)(srcpix + (looph * srcpitch) +
-                                   (1 * MIN(width - 1, loopw + 1)));
-                    H = *(Uint8 *)(srcpix +
-                                   (MIN(height - 1, looph + 1) * srcpitch) +
-                                   (1 * loopw));
-
-                    E0 = D == B && B != F && D != H ? D : E;
-                    E1 = B == F && B != D && F != H ? F : E;
-                    E2 = D == H && D != B && H != F ? D : E;
-                    E3 = H == F && D != H && B != F ? F : E;
-
-                    *(Uint8 *)(dstpix + looph * 2 * dstpitch + loopw * 2 * 1) =
-                        E0;
-                    *(Uint8 *)(dstpix + looph * 2 * dstpitch +
-                               (loopw * 2 + 1) * 1) = E1;
-                    *(Uint8 *)(dstpix + (looph * 2 + 1) * dstpitch +
-                               loopw * 2 * 1) = E2;
-                    *(Uint8 *)(dstpix + (looph * 2 + 1) * dstpitch +
-                               (loopw * 2 + 1) * 1) = E3;
-                }
-            }
-            break;
-        }
-        case 2: {
-            Uint16 E0, E1, E2, E3, B, D, E, F, H;
-            for (looph = 0; looph < height; ++looph) {
-                for (loopw = 0; loopw < width; ++loopw) {
-                    B = *(Uint16 *)(srcpix + (MAX(0, looph - 1) * srcpitch) +
-                                    (2 * loopw));
-                    D = *(Uint16 *)(srcpix + (looph * srcpitch) +
-                                    (2 * MAX(0, loopw - 1)));
-                    E = *(Uint16 *)(srcpix + (looph * srcpitch) + (2 * loopw));
-                    F = *(Uint16 *)(srcpix + (looph * srcpitch) +
-                                    (2 * MIN(width - 1, loopw + 1)));
-                    H = *(Uint16 *)(srcpix +
-                                    (MIN(height - 1, looph + 1) * srcpitch) +
-                                    (2 * loopw));
+    int bytes_ppixel = src->format->BytesPerPixel; 
+    Uint32 E0, E1, E2, E3, B, D, E, F, H;
+    // I had to sacrifice Uint16/8. the compiler shall fill the remaining area with zeroes
```

Comment:
`// I had to sacrifice Uint16/8. the compiler shall fill the remaining area with zeroes`
I'm not sure about the compiler filling the zeros. IIRC you must initialize if manually to the desired bits and bytes.

Also: when it comes to bytes the question comes up: does it still work with little- and/or big-endian platforms?

Reply:
hopefully it should work!